crates/sui-rosetta/src/state.rs

```diff
@@ -0,0 +1,370 @@
+// Copyright (c) 2022, Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+use std::cmp::min;
+use std::collections::BTreeMap;
+use std::str::FromStr;
+use std::sync::Arc;
+use std::time::{Duration, SystemTime, UNIX_EPOCH};
+
+use anyhow::anyhow;
+use async_trait::async_trait;
+use tokio::sync::RwLock;
+use tracing::{debug, error};
+
+use sui_config::genesis::Genesis;
+use sui_core::authority::AuthorityState;
+use sui_core::authority_client::NetworkAuthorityClient;
+use sui_core::quorum_driver::QuorumDriver;
+use sui_types::base_types::{
+    SequenceNumber, SuiAddress, TransactionDigest, TRANSACTION_DIGEST_LENGTH,
+};
+use sui_types::gas_coin::GasCoin;
+use sui_types::object::PastObjectRead;
+
+use crate::operations::Operation;
+use crate::types::{
+    AccountIdentifier, Amount, Block, BlockHash, BlockIdentifier, BlockResponse, CoinAction,
+    CoinChange, CoinID, CoinIdentifier, OperationStatus, OperationType, SignedValue, Transaction,
+    TransactionIdentifier,
+};
+use crate::ErrorType::{BlockNotFound, InternalError};
+use crate::{Error, SUI};
+
+pub struct OnlineServerContext {
+    pub state: Arc<AuthorityState>,
+    pub quorum_driver: Arc<QuorumDriver<NetworkAuthorityClient>>,
+    block_provider: Arc<dyn BlockProvider + Send + Sync>,
+}
+
+impl OnlineServerContext {
+    pub fn new(
+        state: Arc<AuthorityState>,
+        quorum_driver: Arc<QuorumDriver<NetworkAuthorityClient>>,
+        block_provider: Arc<dyn BlockProvider + Send + Sync>,
+    ) -> Self {
+        Self {
+            state,
+            quorum_driver,
+            block_provider,
+        }
+    }
+
+    pub fn blocks(&self) -> &(dyn BlockProvider + Sync + Send) {
+        &*self.block_provider
+    }
+}
+
+#[async_trait]
+pub trait BlockProvider {
+    async fn get_block_by_index(&self, index: u64) -> Result<BlockResponse, Error>;
+    async fn get_block_by_hash(&self, hash: BlockHash) -> Result<BlockResponse, Error>;
+    async fn current_block(&self) -> Result<BlockResponse, Error>;
+    fn genesis_block_identifier(&self) -> BlockIdentifier;
+    async fn oldest_block_identifier(&self) -> Result<BlockIdentifier, Error>;
+    async fn current_block_identifier(&self) -> Result<BlockIdentifier, Error>;
+    async fn get_balance_at_block(
+        &self,
+        addr: SuiAddress,
+        block_height: u64,
+    ) -> Result<u128, Error>;
+}
+
+#[derive(Clone)]
+pub struct PseudoBlockProvider {
+    blocks: Arc<RwLock<Vec<BlockResponse>>>,
+    balance: Arc<RwLock<BTreeMap<SuiAddress, BTreeMap<u64, u128>>>>,
```

Comment:
Can we please document what the keys are here?  What is the inner u64?
What is the plan for recovery of these data structures on failure?

Reply:
good catch, don't need a map here, I am changing it to BTreeMap<SuiAddress, Vec<HistoricalBalance>>